File: apis/src/test/java/com/beat/apis/file/application/FileCommandServiceTest.java
package com.beat.apis.file.application;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.beat.apis.exception.ApiApplicationException;
import com.beat.apis.file.application.command.FileCommandService;
import com.beat.apis.file.exception.FileApplicationErrorCode;
import com.beat.contracts.storage.FileStoragePort;
import com.beat.contracts.storage.PerformancePresignedUrls;
import com.beat.contracts.storage.ImagePresignedUpload;

@ExtendWith(MockitoExtension.class)
class FileCommandServiceTest {

	@Mock
	private FileStoragePort fileStoragePort;

	private FileCommandService fileService;

	@BeforeEach
	void setUp() {
		fileService = new FileCommandService(fileStoragePort);
	}

	@Test
	void issueAllPresignedUrlsNormalizesNullableListsBeforeCallingStoragePort() {
		PerformancePresignedUrls presignedUrls = new PerformancePresignedUrls(
			Map.of("poster", Map.of("poster.png",
				ImagePresignedUpload.of("https://example.com/poster.png", "dev/poster/poster.png")))
		);
		when(fileStoragePort.issueAllPresignedUrlsForPerformanceMaker("poster.png", List.of(), List.of(), List.of()))
			.thenReturn(presignedUrls);

		fileService.issueAllPresignedUrlsForPerformanceMaker(
			"poster.png",
			null,
			null,
			null
		);

		verify(fileStoragePort).issueAllPresignedUrlsForPerformanceMaker("poster.png", List.of(), List.of(), List.of());
	}

	@Test
	void issueAllPresignedUrlsIgnoresLegacyEmptyListPlaceholders() {
		fileService.issueAllPresignedUrlsForPerformanceMaker(
			"poster.png",
			List.of(""),
			List.of(""),
			List.of("")
		);

		verify(fileStoragePort).issueAllPresignedUrlsForPerformanceMaker("poster.png", List.of(), List.of(), List.of());
	}

	@Test
	void issueAllPresignedUrlsRejectsPathLikeFileNamesBeforeCallingStorage() {
		ApiApplicationException exception = assertThrows(
			ApiApplicationException.class,
			() -> fileService.issueAllPresignedUrlsForPerformanceMaker("../poster.png", null, null, null)
		);

		assertEquals(FileApplicationErrorCode.INVALID_FILE_NAME, exception.getErrorCode());
		verifyNoInteractions(fileStoragePort);
	}
}

File: application/frontoffice/src/test/kotlin/com/beat/application/frontoffice/performance/maker/command/FileCommandServiceTest.kt
package com.beat.application.frontoffice.performance.maker.command

import com.beat.application.frontoffice.exception.FrontofficeApplicationException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class FileCommandServiceTest {

    @Mock
    private lateinit var performanceImageStorage: PerformanceImageStorage

    private lateinit var fileService: FileCommandService

    @BeforeEach
    fun setUp() {
        fileService = FileCommandService(performanceImageStorage)
    }

    @Test
    fun `issueAllPresignedUrls normalizes nullable lists before calling storage`() {
        val presignedUrls = PerformancePresignedUrls(
            mapOf(
                "poster" to mapOf(
                    "poster.png" to ImagePresignedUpload.of(
                        "https://example.com/poster.png",
                        "dev/poster/poster.png",
                    ),
                ),
            ),
        )
        Mockito.`when`(
            performanceImageStorage.issueAllPresignedUrls("poster.png", emptyList(), emptyList(), emptyList()),
        ).thenReturn(presignedUrls)

        fileService.issueAllPresignedUrlsForPerformanceMaker("poster.png", null, null, null)

        Mockito.verify(performanceImageStorage).issueAllPresignedUrls(
            "poster.png",
            emptyList(),
            emptyList(),
            emptyList(),
        )
    }

    @Test
    fun `issueAllPresignedUrls ignores legacy empty list placeholders`() {
        fileService.issueAllPresignedUrlsForPerformanceMaker(
            "poster.png",
            listOf(""),
            listOf(""),
            listOf(""),
        )

        Mockito.verify(performanceImageStorage).issueAllPresignedUrls(
            "poster.png",
            emptyList(),
            emptyList(),
            emptyList(),
        )
    }

    @Test
    fun `issueAllPresignedUrls rejects path-like file names before calling storage`() {
        val exception = assertThrows<FrontofficeApplicationException> {
            fileService.issueAllPresignedUrlsForPerformanceMaker("../poster.png", null, null, null)
        }

        assertEquals(FileApplicationErrorCode.INVALID_FILE_NAME, exception.errorCode)
        Mockito.verifyNoInteractions(performanceImageStorage)
    }
}
